platform/linux-generic/odp_ipsec_sad.c

```diff
@@ -900,3 +908,18 @@ uint16_t _odp_ipsec_sa_alloc_ipv4_id(ipsec_sa_t *ipsec_sa)
 	/* No need to convert to BE: ID just should not be duplicated */
 	return tl->next_ipv4_id++;
 }
+
+uint64_t _odp_ipsec_sa_stats_pkts(ipsec_sa_t *sa)
```

Comment:
This does not actually return accurate success stats now because this is based on life time counters and there can be an IPsec processing error  after the life time counter has already been updated. Moving the life time update later in the processing path would not be ok since life time is about how much the SA has been used.

I think you need to track whether an error occurs before or after life time update (e.g. have a new internal counter for that) and adjust the statistics accordingly.

Reply:
For such cases, we can just subtract the 'hard_exp_pkt_err', right?

`+       stats->success = _odp_ipsec_sa_stats_pkts(ipsec_sa) - stats->hard_exp_pkts_err; `

Reply:
You are actually thinking about a different issue. Packets with hard lifetime expiration are should not be counted as succesfuly received packets so something needs to be done about it. Just doing the subtraction like that will not work correctly if byte based life time expires before the packet based.

What I was thinking was the case when lifetime is ok but there is an error later. There are many possibilities for such later error in ipsec_in_single().

Maybe a simple way to fix this could be to add a new flag in the state struct indicating if lifetime counting has already happened and then just before the common error return increment a new internal counter that counts such errored packets. The statistics query could then properly subtract those packets from the success count. Just a quick thought.

By the way, I noticed a problem that lifetime tracking state gets updated before the final anti-replay check, thus potentially allowing replayed packets alter the SA state. See PR 1105 for the fix that touches the same code that we are discussing.

Reply:
Thinking again, do you think we should introduce a different counter for success stats rather than piggybacking on lifetime tracks? Your suggestion looks good otherwise. Let me see how straight forward it is.

Saw the PR. Thanks for fixing it.

Reply:
If it gets very messy piggybacking on lifetime counters then I think we can add a new counter for the success case and optimize things later as necessary.